tests/Helpers/JWKFetcherTest.php

```diff
@@ -0,0 +1,195 @@
+        $this->assertEquals( '__test_x5c_1__', $pem_parts[1] );
+    }
+
+    /**
+     * Test that a standard JWKS path returns the correct x5c when given a kid.
+     *
+     * @return void
+     */
+    public function testGetJwksX5cWithKid()
+    {
+        $jwksFetcher = $this->getStub();
+
+        $pem = $jwksFetcher->getJwksX5c( 'https://localhost/.well-known/jwks.json', '__test_kid_2__' );
+
+        $this->assertNotEmpty($pem);
+
+        $pem_parts = explode( PHP_EOL, $pem );
+
+        $this->assertEquals( 4, count($pem_parts) );
+        $this->assertEquals( '-----BEGIN CERTIFICATE-----', $pem_parts[0] );
+        $this->assertEquals( '-----END CERTIFICATE-----', $pem_parts[2] );
+        $this->assertEquals( '__test_x5c_2__', $pem_parts[1] );
+    }
+
+    /**
+     * Test that a custom JWKS path returns the correct JSON.
+     *
+     * @return void
+     */
+    public function testGetJwksX5cWithPath()
+    {
+        $jwksFetcher = $this->getStub();
+
+        $pem = $jwksFetcher->getJwksX5c( 'https://localhost/.custom/jwks.json', '__test_custom_kid__' );
+
+        $this->assertNotEmpty($pem);
+
+        $pem_parts = explode( PHP_EOL, $pem );
+
+        $this->assertEquals( 4, count( $pem_parts ) );
+        $this->assertEquals( '-----BEGIN CERTIFICATE-----', $pem_parts[0] );
+        $this->assertEquals( '-----END CERTIFICATE-----', $pem_parts[2] );
+        $this->assertEquals( '__test_custom_x5c__', $pem_parts[1] );
+    }
+
+    /**
+     * Test that a JWK with no x509 cert returns a blank array.
+     *
+     * @return void
+     */
+    public function testGetJwksX5cNoX5c()
+    {
+        $jwksFetcher = $this->getStub();
+
+        $pem = $jwksFetcher->getJwksX5c( 'https://localhost/.custom/jwks.json', '__no_x5c_test_kid_2__' );
+
+        $this->assertEmpty($pem);
+    }
+
+    /**
+     * Test that the protected getProp method returns correctly.
+     *
+     * @return void
+     */
+    public function testConvertCertToPem()
+    {
+        $class  = new \ReflectionClass(JWKFetcher::class);
+        $method = $class->getMethod('convertCertToPem');
+        $method->setAccessible(true);
+
+        $test_string_1 = '';
+        for ($i = 1; $i <= 64; $i++) {
+            $test_string_1 .= 'a';
+        }
+
+        $test_string_2 = '';
+        for ($i = 1; $i <= 64; $i++) {
+            $test_string_2 .= 'b';
+        }
+
+        $returned_pem = $method->invoke(new JWKFetcher(), $test_string_1.$test_string_2);
+        $pem_parts    = explode( PHP_EOL, $returned_pem );
+        $this->assertEquals( 5, count($pem_parts) );
+        $this->assertEquals( '-----BEGIN CERTIFICATE-----', $pem_parts[0] );
+        $this->assertEquals( $test_string_1, $pem_parts[1] );
+        $this->assertEquals( $test_string_2, $pem_parts[2] );
+        $this->assertEquals( '-----END CERTIFICATE-----', $pem_parts[3] );
+    }
+
+    public function testCacheReturn()
+    {
+        $cache_handler = new TestCacheHandler();
```

Comment:
Testing the CacheHandler this way is sort of wrong, since this `TestCacheHandler` is not the implementation that the user will use, right? *wink*. Instead, you should mock a `CacheHandler` and assert that `get/set/remove` methods get called N times with the given XYZ arguments. You can then remove the `TestCacheHandler` class introduced in this PR.

Reply:
Well, we don't know the implementation being used, right? So, in this case, it's testing that "some" CacheHandler implementation will work, as long as it's following the right interface. I could add this as a `MemoryCacheHandler` "official" implementation?

Reply:
> Well, we don't know the implementation being used, right?
No, but you know the interface and you know you're going to call it to 
a) store
b) retrieve
c) delete

Adding a subclass for testing purposes is not wrong but it's not the right way either. IMO, you should create a new mock from the interface definition and assert the calls are being made correctly. 

>I could add this as a MemoryCacheHandler "official" implementation
If you want, go ahead. But remember whatever you ship can't be undone 💃 and unless you have a "best practice in storage/cache" to recommend users to begin using this class, I'd stay away